### PR TITLE
Add text classification support

### PR DESCRIPTION
**Pull request by jinyichao**
Hi Davis, as discussed, I simply use the average feature vectors of each word as the feature vector for the entire document. Overall, there are a few new files to support this new feature, but not any single line deletion on existing codes. Therefore, all the existing features should not be affected, and future improvements are expected based on these new files.  

Regarding the performance in its current form, I test it based on a small internal mail classification dataset, the result is surprisingly good (e.g., more than 90% F1-score). 

I would really appreciate if you can spare some time to review the code, and give me your valuable comments. Thanks.

### CODE REVIEW

**Comment by davisking** [comment]
Cool.  I'll look at this and let you know.

You should also add a simple bag-of-words feature vector into it using [feature hashing](https://en.wikipedia.org/wiki/Feature_hashing).  It's really easy to implement.  I've often used dlib's [murmur_hash3_128bit](http://dlib.net/dlib/general_hash/murmur_hash3_abstract.h.html#murmur_hash3_128bit) function to implement the second version of the "hashing_vectorizer" function shown in that wikipeida page and it works quite well.  The cross validation code should also try just using the "total word features", the hashing features, and a combination concatenated together and pick whatever works best.  That would make this feature reasonably complete for a first version.

**Comment by jinyichao** [comment]
OK, thanks for your detailed instructions. Previously, I was afraid that using bag-of-words alone cannot efficiently use the "word features", thus is not expected to work well for a small set of test cases. But yeah, using a combination of them would be much better. Will try to add the implementation soon.

**Comment by davisking** [comment]
Yeah, it definitely won't work by itself for small datasets.

**Comment by jinyichao** [comment]
I tested both the "total word feature" and "bag-of-words hashing features" based text classification on a public [20newsgroups dataset] (http://qwone.com/~jason/20Newsgroups/). In particular, I simplify the task a bit by only focusing on the 4 coarse types (i.e., comp, rec, sci, and talk), as shown in the table, and uses exactly the train and test files inside the 20news-bydate.tar.gz download from that website without any text pre-processing.  

"total word features" alone achieves 88.7% F1-score, while "bag-of-words" alone only achieves 72.7% F1-score. This result actually matches well with my expectation. Because intuitively, "total word feature" is trained from a large and high-quality corpus, making it able to properly represent the meaning of each word in some level. Whereas the "bag-of-words" only learns the word occurrence from the training data, although, its poor performance might be also attributed to the small hash vector size (e.g., 10000) in the code and other implementation issues. 

And for the combined solution, I simply append the "bag-of-words" hashing vectors to the "total word features", to make a larger feature_vector for the text. In this way, it gives almost the same F1-score (e.g., 88.8%) for the same task.

**Comment by davisking** [comment]
Cool.

It's been a while since I looked at this code.  So I forgot that MITIE
already uses these kinds of hashed features in the ner labeling code.  In
particular, make_feat() creates the features you need.  I see that you
found it but you shouldn't duplicate code by copying it into another file.
Code should be shared by use of header files.  Also, the
way extract_BoW_features works isn't quite right.  It should look
like extract_text_features except without the "total word features".  I
also expect BoW features to work a little better on largish training sets
and 20 newsgroups seems big enough.  So this might be the reason.

**Comment by jinyichao** [comment]
I guess you want to say, 20newsgroup is NOT big enough. :) if in this case, yes, I agree :)

And thanks for pointing out the problems in the code. I will modify them accordingly and let you know later.

**Comment by jinyichao** [comment]
Hi Davis, I just realized that result.push_back(make_feat(shash(words[i],0))) and result.push_back(make_feat(shash(stem_word(words[i]),10))) actually achieves the bag-of-words features. So I made some modifications, and now, "word features" alone achieves 74.9% F1-score, BoW alone is 87.3%, and the combined solution is still 88.8%, for the 20newsgroups task.

**Comment by davisking** [comment]
Cool, that's more the F1 scores I would expect for those methods :)

Looking through the code I see a few issues still remaining:

The name of the program isn't right here (https://github.com/mit-nlp/MITIE/pull/49/files#diff-5b16cbdea5f182df2cd1d6cf206784bbR68). 

text_categorizer_extractor should be renamed to text_categorizer.  Especially since that matches the name in the Java API.

Is this comment correct? https://github.com/mit-nlp/MITIE/pull/49/files#diff-5b16cbdea5f182df2cd1d6cf206784bbR85

I see some cases of copy and pasted code that doesn't actually do anything. e.g. https://github.com/mit-nlp/MITIE/pull/49/files#diff-fd32f4cff3a5bcf0ef9f7a6bc66d05a7R17

The requires/ensures comments for this function don't make sense: https://github.com/mit-nlp/MITIE/pull/49/files#diff-3684754a9aae1aaf1ea650c305c64f89R41

The usage and behavior of the "Unseen" label should be clarified in the documentation.

Fix tabbing:https://github.com/mit-nlp/MITIE/pull/49/files#diff-2cbf13001db8c576b703ef9d81a735cdR124

There should also be an easy way to use just BoW features since that allows the user to use much smaller models.  The total word feature extractor is very large, I'm sure many users won't care about the additional accuracy it gives for text categorization considering the computational requirements it imposes.

**Comment by jinyichao** [comment]
Thanks for the detailed comments! According to them, I made two more commits, please kindly check them. :)

**Comment by davisking** [comment]
Cool.  I'll look at this later this weekend.

**Comment by davisking** [comment]
Been crazy busy the last few days, so I didn't get to this.  I'll do it
this week at some point :)

**Comment by jinyichao** [comment]
No hurry! Take your time. :)

**Comment by davisking** [comment]
I just pulled the code and ran `make test` at the top and got an error.  Can you make sure the tests run?

**Comment by jinyichao** [comment]
Oops, I forgot to check "make test" command before, and there were some minor issues in the makefile. They should be fixed now, please check them one more time. :)

**Comment by davisking** [comment]
Cool.  Now `make test` works but `make clean; make` gives an error.

**Comment by jinyichao** [comment]
Err...... sorry. I forgot to include the text categorizer into the example list. It is there now. Eventually, I hope it should work now...

**Comment by davisking** [comment]
Super, this is looking pretty good.  There are just a few more minor things:

If operator() outputs just one argument then it should return it using the `return` keyword rather than via a named parameter: https://github.com/mit-nlp/MITIE/pull/49/files#diff-d8e0976856497c45f7a97bddbf09ab1bR83

The text_categorizer_BoW example should be renamed to something like train_text_categorizer_BoW so that it's name is cohesive with the other mode of training and other examples.  That is, this way it is clear that there are two ways to train a text categorizer and that this example is one of them.

Don't put `using namespace` declarations in header files: https://github.com/mit-nlp/MITIE/pull/49/files#diff-eea8b6e1f1ecf0f8fe1b869bbed5f3c6R14

I went to `MITIE/mitielib/java/build` and ran `cmake ..; make install` and got this error
```
In file included from /home/davis/MITIE/mitielib/java/build/lib/java_src/edu/mit/ll/mitie/swigJAVA_wrap.cxx:231:0:
/home/davis/MITIE/mitielib/java/./swig_api.h:28:30: fatal error: text_categorizer.h: No such file or directory
 #include "text_categorizer.h"
```


This line should say 'document tag', not NER tag https://github.com/mit-nlp/MITIE/pull/49/files#diff-1d050bd526db4a845d0c51a9ff6a75aaR58

**Comment by jinyichao** [comment]
Hi, Davis, Thanks for your careful check and detailed comments!

I've changed the code accordingly, please spare some of your time to go through it again. :)

**Comment by davisking** [comment]
Cool.  Almost there :)

I see a bunch of places that talk about the `text_categorizer_extractor` but that doesn't exist anymore.  Can you go through the comments and make sure everything is up to date?

Also, can you rename the `train_categorizer` example to `train_text_categorizer` so that it matches the naming scheme of the other examples i.e. some_example, train_some_example.

**Comment by jinyichao** [comment]
Ha, thanks~ I renamed them through the whole project. Just check it :)

**Comment by davisking** [comment]
Cool.  However, looking over it again, I'm still finding problems.  For example.  in the cpp/train_text_categorizer, the message it prints when it is run is, 

> You must give the path to the MITIE English total_word_feature_extractor.dat file.
> So run this program with a command like: 
> ./train_categorizer_example ../../../MITIE-models/english/total_word_feature_extractor.dat

Which is the wrong command.  

I think the PR is about ready to go in terms of the code and features, but these little details need to be taken care of.  Can you read over the whole PR and make sure the documentation, comments, and so forth is correct?  That's what I'm doing each time I go over a PR and when I start to find these kinds of things I stop and send it back to you.  The fewer times I have to read over the PR the better, and the sooner it will be merged :)

**Comment by jinyichao** [comment]
Hi, Davis, I really appreciate your patient and careful review! It is really a good lesson for me to how to contribute to a high-quality project! 

Just going through the PR again, and revised a few places. Please do not hesitate to point out if there is any further problem. :)

**Comment by davisking** [comment]
Super, thanks for the PR.  I just merged it into the main repository and fixed few minor remaining details.

**Comment by jinyichao** [comment]
Really a great news for me! As I said, it is also a very good lesson for me! Thanks a lot!

**Comment by davisking** [comment]
No problem :)
